core/trino-main/src/main/java/io/trino/sql/routine/SqlRoutinePlanner.java

```diff
@@ -0,0 +1,503 @@
+            RowExpression rowExpression = translator.process(optimized, null);
+
+            // optimize RowExpression
+            ExpressionOptimizer optimizer = new ExpressionOptimizer(plannerContext.getMetadata(), plannerContext.getFunctionManager(), session);
+            rowExpression = optimizer.optimize(rowExpression);
+
+            return rowExpression;
+        }
+
+        private static Map<NodeRef<LambdaArgumentDeclaration>, Symbol> buildLambdaDeclarationToSymbolMap(Analysis analysis, SymbolAllocator symbolAllocator)
+        {
+            Map<Key, Symbol> allocations = new HashMap<>();
+            Map<NodeRef<LambdaArgumentDeclaration>, Symbol> result = new LinkedHashMap<>();
+
+            for (Map.Entry<NodeRef<Expression>, Type> entry : analysis.getTypes().entrySet()) {
+                if (!(entry.getKey().getNode() instanceof LambdaArgumentDeclaration argument)) {
+                    continue;
+                }
+
+                Key key = new Key(argument, entry.getValue());
+
+                // Allocate the same symbol for all lambda argument names with a given type. This is needed to be able to
+                // properly identify multiple instances of syntactically equal lambda expressions during planning as expressions
+                // get rewritten via TranslationMap
+                Symbol symbol = allocations.get(key);
+                if (symbol == null) {
+                    symbol = symbolAllocator.newSymbol(argument, entry.getValue());
+                    allocations.put(key, symbol);
+                }
+
+                result.put(NodeRef.of(argument), symbol);
+            }
+
+            return result;
+        }
+
+        private record Key(LambdaArgumentDeclaration argument, Type type)
+        {
+            private Key
+            {
+                requireNonNull(argument, "argument is null");
+                requireNonNull(type, "type is null");
+            }
+        }
+
+        public static Expression coerceIfNecessary(Analysis analysis, Expression original, Expression rewritten)
+        {
+            Type coercion = analysis.getCoercion(original);
+            if (coercion == null) {
+                return rewritten;
+            }
+            return new Cast(rewritten, toSqlType(coercion), false, analysis.isTypeOnlyCoercion(original));
+        }
+
+        private ExpressionAnalyzer createExpressionAnalyzer(Session session, TypeProvider typeProvider)
+        {
+            return ExpressionAnalyzer.createWithoutSubqueries(
+                    plannerContext,
+                    new AllowAllAccessControl(),
+                    session,
+                    typeProvider,
+                    ImmutableMap.of(),
+                    node -> semanticException(NOT_SUPPORTED, node, "Queries are not allowed in functions"),
```

Comment:
This should be caught during analysis. At this point, it can just be some IllegalXXXException, since it should never occur based on user input.